opentelemetry-sdk/src/trace/span.rs

```diff
@@ -725,4 +732,70 @@ mod tests {
         // return none if the provider has already been dropped
         assert!(dropped_span.exported_data().is_none());
     }
+
+    #[test]
+    #[cfg(feature = "experimental_span_processor_on_ending")]
+    fn test_on_ending_mutate_span() {
+        use crate::trace::SpanProcessor;
+
+        #[derive(Debug)]
+        struct FirstProcessor;
+        impl SpanProcessor for FirstProcessor {
+            fn on_start(&self, _span: &mut Span, _cx: &opentelemetry::Context) {}
+            fn on_end(&self, _span: crate::trace::SpanData) {}
+            fn on_ending(&self, span: &mut Span) {
+                span.set_attribute(KeyValue::new("first_processor", "true"));
+            }
+
+            fn force_flush(&self) -> crate::error::OTelSdkResult {
+                Ok(())
+            }
+
+            fn shutdown_with_timeout(&self, _timeout: Duration) -> crate::error::OTelSdkResult {
+                Ok(())
+            }
+        }
+
+        #[derive(Debug)]
+        struct SecondProcessor;
+        impl SpanProcessor for SecondProcessor {
+            fn on_start(&self, _span: &mut Span, _cx: &opentelemetry::Context) {}
+            fn on_end(&self, _span: crate::trace::SpanData) {}
+            fn on_ending(&self, span: &mut Span) {
+                assert!(span
+                    .exported_data()
+                    .unwrap()
+                    .attributes
+                    .contains(&KeyValue::new("first_processor", "true")));
+                span.set_attribute(KeyValue::new("second_processor", "true"));
+            }
+
+            fn force_flush(&self) -> crate::error::OTelSdkResult {
+                Ok(())
+            }
+
+            fn shutdown_with_timeout(&self, _timeout: Duration) -> crate::error::OTelSdkResult {
+                Ok(())
+            }
+        }
+
+        let exporter = crate::trace::span_processor::tests::MockSpanExporter::new();
+        let spans = exporter.exported_spans.clone();
+
+        let provider = crate::trace::SdkTracerProvider::builder()
+            .with_span_processor(FirstProcessor)
+            .with_span_processor(SecondProcessor)
+            .with_simple_exporter(exporter)
+            .build();
+        provider.tracer("test").start("test_span");
```

Comment:
test suggestion:
start the span with some attributes.
then add few more attributes to it.
validate that on_start sees only those attributes added at start time.
validate that on_ending sees attributes added at start time, attributes added via span.setattribute, attributes added by any previous processor.